src/Interfaces.jl

```diff
@@ -202,19 +245,29 @@ macro interface(IT, alias_or_block, maybe_block=nothing)
     return esc(quote
         Interfaces.isinterfacetype(::Type{$IT}) = true
         Interfaces.interface(::Type{$IT}) = $iface
-        function Interfaces.implements(::Type{T}, ::Type{$IT}, mods::Vector{Module}=[parentmodule(T)]) where {T}
+        function Interfaces.implements(::Type{T}, ::Type{$IT}, mods::Vector{Module}=[parentmodule(T)]; debug::Bool=false) where {T}
+            debuglvl = debug ? :error : debug
             $block
         end
     end)
 end
 
-macro implements(T, IT)
+macro implements(T, IT, debug=false)
+    x = debug isa Bool ? debug :
+        (debug isa Expr && debug.head == :(=)) ? debug.args[2] :
+        throw(ArgumentError("unsupported argument in @implements: `$debug`"))
     return esc(quote
-        @assert Interfaces.implements($T, $IT)
+        @assert Interfaces.implements($T, $IT; debug=$x)
         Interfaces.implements(::Type{$T}, ::Type{$IT}) = true
     end)
 end
 
-@noinline returntype(@nospecialize(f), @nospecialize(args)) = Base.return_types(f, args)[1]
+struct NoMethod end
+
+@noinline function returntype(@nospecialize(f), @nospecialize(args))
+    rt = Base.return_types(f, args)
+    # Avoid BoundsError in debug mode when no method found
+    return isempty(rt) ? NoMethod : rt[1]
```

Comment:
I wasn't sure what would be best here

without this we get a BoundsError which i found pretty inscrutable 
```jl
julia> Interfaces.implements(Foo, AbstractArray, [Base]; debug=true)
[ InterfaceImplementationError: missing `AbstractArray` interface method definition: `(IndexStyle)(::Type{Foo})`, in module(s): `Module[Base]`
ERROR: BoundsError: attempt to access 0-element Vector{Any} at index [1]
Stacktrace:
 [1] getindex
   @ ./array.jl:801 [inlined]
 [2] returntype(f::Any, args::Any)
   @ Interfaces ~/repos/Interfaces.jl/src/Interfaces.jl:271
 [3] implements(::Type{Foo}, ::Type{AbstractArray}, mods::Vector{Module}; debug::Bool)
   @ Main ~/repos/Interfaces.jl/src/Interfaces.jl:111
```
whereas now we would get a MethodError (due to actually missing a require method), which seems much more reasonable to me:
```jl
julia> Interfaces.implements(Foo, AbstractArray, [Base]; debug=true)
[ InterfaceImplementationError: missing `AbstractArray` interface method definition: `(IndexStyle)(::Type{Foo})`, in module(s): `Module[Base]`
[ InterfaceImplementationError: invalid return type for `AbstractArray` interface method definition: `(IndexStyle)(::Type{Foo})`; inferred Interfaces.NoMethod, required Union{IndexCartesian, IndexLinear}
ERROR: MethodError: no method matching IndexStyle(::Type{Foo})
Closest candidates are:
  IndexStyle(::Base.IndexSCartesian2{K}, ::Base.IndexSCartesian2{K}) where K at reinterpretarray.jl:198
  IndexStyle(::Base.Broadcast.Broadcasted) at broadcast.jl:226
  IndexStyle(::SuiteSparse.CHOLMOD.Sparse) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/SuiteSparse/src/cholmod.jl:1191
  ...
Stacktrace:
 [1] implements(::Type{Foo}, ::Type{AbstractArray}, mods::Vector{Module}; debug::Bool)
   @ Main ~/repos/Interfaces.jl/src/Interfaces.jl:115
```
I think this is fine... i don't think `debug=true` needs to promise that no errors will be thrown (or if it should, that might reequire a bit more thinking!)

I thought about what `NoMethod` here should be. I considered `Union{}`, but then the subtyping check (`Union{} <: RT`) would alway pass, which doesn't seem right... but any other answer than `Union{}` (e.g. `Any`) just feels wrong... so I went for a custom type that will always fail the subtyping check and hopefully be clear enough in the resulting message (show above)

Reply:
changed to just acoid this code path if there's no method